### PR TITLE
FED-3143 Null Safety Codemod for connect props

### LINKED ISSUE #295: Add codemod for connect and null safety
Currently, after migrating connected components, required prop validation throws when consuming those components for any required props that are only set within mapStateToProps/mapActionsToProps/etc.

See: https://github.com/Workiva/over_react/issues/942#issuecomment-2353692089

We should add a codemod to automatically disable validation for these props.

### PR DESCRIPTION
**Pull request by sydneyjodon-wk**
This is based on #298 to share the new e2e tests

Closes #295

## Motivation
  
Currently, after migrating connected components, required prop validation throws when consuming those components for any required props that are only set within mapStateToProps/mapActionsToProps/etc.

https://workiva.slack.com/archives/CEFTMBPAA/p1726516120035799
https://github.com/Workiva/over_react/issues/942#issuecomment-2353692089
There are quite a few usages of connect ([SourceGraph](https://workiva.sourcegraphcloud.com/search?q=context:global+-repo:over_react+select:file+lang:dart+/%5Cbconnect%28Flux%29%3F%5Cb/+and+/map%5Cw%2BToProps/+&patternType=keyword&sm=0)) that will be impacted as they migrate to null safety, so we should codemod them

## Changes
  
- Add codemod to add disabled prop validation annotations for connect usages
## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

Please review: 

### QA Checklist
- [ ] Tests were updated and provide good coverage of the changeset and other affected code
- [ ] Manual testing was performed if needed
    - [ ] Steps from PR author: 
        
      - [ ] CI passes
      - [ ] Good test coverage
      - [ ] QA in another repo
    - [ ] Anything falling under manual testing criteria [outlined in CONTRIBUTING.md][contributing-manual-testing]

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:
- [ ] A Frontend Frameworks Design member has reviewed these changes
- [ ] There are no unaddressed comments _- this check can be automated if reviewers use the "Request Changes" feature_
- [ ] _For release PRs -_ Version metadata in Rosie comment is correct

[contributing-review-types]: https://github.com/Workiva/over_react_codemod/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/over_react_codemod/blob/master/CONTRIBUTING.md#manual-testing-criteria

### CODE REVIEW

**Review by aaronlademann-wf — APPROVED** [review]

File: test/test_fixtures/required_props/test_package/lib/src/test_state.dart [hunk 1]
```diff
@@ -1,15 +1,18 @@
...
+import 'package:over_react/over_react_redux.dart';
 
 // ignore: uri_has_not_been_generated
 part 'test_state.over_react.g.dart';
 
-UiFactory<FooProps> Foo =
-    castUiFactory(_$Foo); // ignore: undefined_identifier
+UiFactory<FooProps> Foo = connect<FooState, FooProps>(
+  mapStateToPropsWithOwnProps: (state, props) => Foo()..prop1 = 1,
+)(castUiFactory(_$Hoc)); // ignore: undefined_identifier
```
Comment (greglittlefield-wf) [inline test/test_fixtures/required_props/test_package/lib/src/test_state.dart:11]: #nit I think this should be:
```suggestion
)(castUiFactory(_$Foo)); // ignore: undefined_identifier
```

File: lib/src/dart3_suggestors/null_safety_prep/connect_required_props.dart [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    if (result == null) {
+      throw Exception(
+          'Could not get resolved result for "${context.relativePath}"');
+    }
+    result.unit.accept(this);
+
+    // Add the patches at the end so that all the props to be ignored can be collected
+    // from the different args in `connect` before adding patches to avoid duplicate patches.
+    for (final propsClass in _ignoredPropsByMixin.keys) {
+      final propsToIgnore = _ignoredPropsByMixin[propsClass]!;
```
Comment (greglittlefield-wf) [inline lib/src/dart3_suggestors/null_safety_prep/connect_required_props.dart:83]: #nit In general, when iterating over Map key/value pairs, I think it's a good idea to use null-safe APIs, such as `.forEach` or iterating over `.entries`. 

That way, there's no risk of `!` being incorrectly applied to value if it's nullable (e.g., if this map was typed as Map<InterfaceElement, List<String>?>`).

**Review by greglittlefield-wf — APPROVED** [review]
+1, QAing now

File: lib/src/dart3_suggestors/null_safety_prep/connect_required_props.dart [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+      // Keep a running list of props to ignore per props mixin.
+      final fieldName = field.name.name;
+      if (_ignoredPropsByMixin[propsElement] != null) {
+        if (!_ignoredPropsByMixin[propsElement]!.contains(fieldName)) {
+          _ignoredPropsByMixin[propsElement]!.add(fieldName);
+        }
+      } else {
+        _ignoredPropsByMixin[propsElement] = [fieldName];
+      }
```
[the remark this answers is not in the document]
Comment (sydneyjodon-wk) [reply lib/src/dart3_suggestors/null_safety_prep/connect_required_props.dart:66]: Wow that's beautiful - thank you!

**Review by greglittlefield-wf — APPROVED** [review]
+10

@Workiva/release-management-p

**Review by rmconsole-wf — APPROVED** [review]
+1 from RM

**Pull request merged**